agent/xds/jwt_authn.go

```diff
@@ -0,0 +1,249 @@
+package xds
+
+import (
+	"encoding/base64"
+	"fmt"
+
+	envoy_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoy_route_v3 "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	envoy_http_jwt_authn_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/jwt_authn/v3"
+	envoy_http_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	"github.com/hashicorp/consul/agent/structs"
+	"google.golang.org/protobuf/types/known/durationpb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+func makeJWTAuthFilter(pCE map[string]*structs.JWTProviderConfigEntry, intentions structs.SimplifiedIntentions) (*envoy_http_v3.HttpFilter, error) {
+	providers := map[string]*envoy_http_jwt_authn_v3.JwtProvider{}
+	var rules []*envoy_http_jwt_authn_v3.RequirementRule
+
+	for _, intention := range intentions {
+		for _, jwtReq := range collectJWTRequirements(intention) {
+			if _, ok := providers[jwtReq.Name]; ok {
+				continue
+			}
+
+			jwtProvider, ok := pCE[jwtReq.Name]
+
+			if !ok {
+				return nil, fmt.Errorf("provider specified in intention does not exist. Provider name: %s", jwtReq.Name)
+			}
+			envoyCfg, err := buildJWTProviderConfig(jwtProvider)
+			if err != nil {
+				return nil, err
+			}
+			providers[jwtReq.Name] = envoyCfg
+		}
+	}
```

Comment:
I was wondering why you loop through `intentions` twice in this function. It looks like the first loop is for building a set of `providers` and the second is for extracting rules. 

Would it be possible to do both in one loop?

It might get a little messy but the `xds` package should prioritize performance since a lot of the codepaths are hot.